src/gserde.gleam

```diff
@@ -35,15 +38,99 @@ fn to_output_filename(src_filename) {
 }
 
 pub fn main() {
-  let path = case argv.load().arguments {
-    [path] -> path
-    _ -> panic as "path to a directory is required as a first argument"
+  // let path = case argv.load().arguments {
+  //   [path] -> path
+  //   _ -> panic as "path to a directory is required as a first argument"
+  // }
+
+  // let filter = filter.Everything
+
+  // process_path(filter, path)
+
+  let params =
+    config.Params(
+      project_root: ".",
+      subpaths_filter: filter.Dirs(["src"]),
+      gen_filter: filter.Everything,
+      serde_filter: filter.Both,
+    )
+
+  process(params, True)
+}
+
+pub fn inspect(data: t, msg: String, do: Bool, f) -> t {
+  case do {
+    False -> data
+    True -> {
+      io.println(msg)
+      io.println(string.inspect(f(data)))
+
+      data
+    }
+  }
+}
+
+pub fn debug_inspect(data: t, msg: String, do_print: Bool) -> t {
+  inspect(data, msg, do_print, function.identity)
+}
+
+fn filter_by_directories(
+  paths: List(String),
+  dirs: List(String),
+) -> List(String) {
+  list.filter(paths, fn(path) {
+    list.any(dirs, fn(dir) {
+      // FIXME: holy shit fix this matching
+      string.starts_with(path, dir <> "/")
+      || string.starts_with(path, "./" <> dir <> "/")
+      || path == dir
+      || path == "./" <> dir
+    })
+  })
+}
+
+fn find_gleam_files(root: String) -> List(String) {
+  fswalk.builder()
+  |> fswalk.with_path(root)
+  // |> fswalk.with_traversal_filter(fn(it) {
+  //   string.ends_with(it.filename, ".gleam") && !it.stat.is_directory
+  // })
+  |> fswalk.walk
+  |> yielder.map(fn(v) { expect(v, "failed to walk").filename })
+  |> yielder.filter(fn(v) {
+    // I have no idea why the traversal filter still includes directories.
+    string.ends_with(v, ".gleam")
+  })
+  |> yielder.to_list
+}
+
+pub fn filter_files(paths: List(String), filter: filter.SubpathsFilter) {
+  case filter {
+    filter.All -> paths
+    filter.Dirs(dirs) -> filter_by_directories(paths, dirs)
   }
+}
+
+pub fn process(params: config.Params, verbose: Bool) {
+  debug_inspect(params, "Parameters:", verbose)
 
-  process_path(path)
+  params.project_root
+  |> find_gleam_files()
+  // |> debug_inspect("All found files:", verbose)
+  // TODO: load files according to subpaths filter
+  |> filter_files(params.subpaths_filter)
+  |> debug_inspect("Filtered files:", verbose)
+
+  // parse files, aggregate into a map with filepath, module path, parsed tree
+
+  // filter out files based on TypeFilter
+
+  //
```

Comment:
also: TODO generated data, and then return a map of `generated file path -> contents`

Reply:
more or less done